File: packages/shared-types/bot.ts
import type { IsoDatetimeString, JsonObject } from './json';

export enum BotStatus {
  DRAFT = 'draft',
  PAPER_ACTIVE = 'paper_active',
  LIVE_PENDING_APPROVAL = 'live_pending_approval',
  LIVE_ACTIVE = 'live_active',
  PAUSED = 'paused',
  STOPPED = 'stopped',
  ERROR = 'error',
}

export enum BotMode {
  PAPER = 'paper',
  LIVE = 'live',
}

export interface Bot {
  id: string;
  userId: string;
  exchangeAccountId?: string;
  name: string;
  mode: BotMode;
  strategy: string;
  symbol: string;
  quoteAsset: string;
  capital: number;
  status: BotStatus;
  config: JsonObject;
  createdAt: Date;
  updatedAt: Date;
}

export interface BotWire {
  id: string;
  user_id: string;
  exchange_account_id?: string;
  name: string;
  mode: BotMode;
  strategy: string;
  symbol: string;
  quote_asset: string;
  capital: number;
  status: BotStatus;
  config: JsonObject;
  created_at: IsoDatetimeString;
  updated_at: IsoDatetimeString;
}

export interface RiskSettings {
  id: string;
  botId: string;
  maxPositionSize?: number;
  maxDailyLoss?: number;
  stopLossPercent?: number;
  takeProfitPercent?: number;
  emergencyStop: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export interface RiskSettingsWire {
  id: string;
  bot_id: string;
  max_position_size?: number;
  max_daily_loss?: number;
  stop_loss_percent?: number;
  take_profit_percent?: number;
  emergency_stop: boolean;
  created_at: IsoDatetimeString;
  updated_at: IsoDatetimeString;
}

File: services/executor-rust/src/types/bot.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, Type};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone, Type)]
#[serde(rename_all = "snake_case")]
#[sqlx(rename_all = "snake_case", type_name = "varchar")]
pub enum BotStatus {
    Draft,
    PaperActive,
    LivePendingApproval,
    LiveActive,
    Paused,
    Stopped,
    Error,
}

#[derive(Debug, Serialize, Deserialize, Clone, Type)]
#[serde(rename_all = "snake_case")]
#[sqlx(rename_all = "snake_case", type_name = "varchar")]
pub enum BotMode {
    Paper,
    Live,
}

#[derive(Debug, Serialize, Deserialize, Clone, FromRow)]
pub struct Bot {
    pub id: uuid::Uuid,
    pub user_id: uuid::Uuid,
    pub exchange_account_id: Option<uuid::Uuid>,
    pub exchange: Option<String>, // Added field
    pub name: String,
    pub mode: BotMode,
    pub strategy: String,
    pub symbol: String,
    pub quote_asset: String,
    pub capital: f64,
    pub status: BotStatus,
    pub config: sqlx::types::Json<HashMap<String, serde_json::Value>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, Clone, FromRow)]
pub struct RiskSettings {
    pub id: uuid::Uuid,
    pub bot_id: uuid::Uuid,
    pub max_position_size: Option<f64>,
    pub max_daily_loss: Option<f64>,
    pub max_drawdown_percent: Option<f64>,
    pub stop_loss_percent: Option<f64>,
    pub take_profit_percent: Option<f64>,
    pub trailing_stop_percent: Option<f64>,
    pub emergency_stop: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
